obspy/io/rg16/util.py

```diff
@@ -39,131 +35,105 @@ def _open_file(func, *args, **kwargs):
             raise  # type error was in function call, not in opening file
         out = func(*args, **kwargs)
         first_arg.seek(0)  # reset position to start of file
-        return out
-
-
-READ_FUNCS = {}
-
-
-def _register_read_func(dtype):
-    def _wrap(func):
-        READ_FUNCS[dtype] = func
-        return func
-
-    return _wrap
-
-
-def _read_block(fi, spec, start_bit=0):
-    out = {}
-    for name, start, length, fmt in spec:
-        out[name] = _read(fi, start_bit + np.array(start), length, fmt)
     return out
 
 
-def _read(fi, position, length, dtype, new_dtype=None):
+def _read(fi, position, length, dtype, left_part=True):
     """
     Read one or more bytes using provided datatype.
 
     :param fi: A buffer containing the bytes to read.
     :param position: Byte position to start reading.
     :type position: int
     :param length: Length, in bytes, of data to read.
-    :type length: int
-    :param dtype:
-        The data type, all numpy data types are supported plus the following:
-            bcd - binary coded decimal
-            <i3 - little endian 3 byte int
-            >i3 - big endian 3 byte int
-            >i. - 4 bit int, left four bits
+    :type length: int or float
+    :param dtype: bcd, binary or IEEE
     :type dtype: str
-    :param new_dtype: Any valid numpy data type.
-    :return:
+    :param left_part: If True, start the reading from the first half part
+        of the byte position. If False, start the reading from the second
+        half part of the byte position.
+    :type left_part: boolean
     """
-    # if a list is passed as parameters then recurse through each
-    if isinstance(position, (collections.Sequence, np.ndarray)):
-        assert len(position) == len(length) == len(dtype)
-        for pos, leng, dty in zip(position, length, dtype):
-            try:
-                return _read(fi, pos, leng, dty)
-            except ValueError:
-                pass
-        else:
-            msg = 'failed to read chunk'
-            raise ValueError(msg)
-    # non recursive case
     fi.seek(position)
-    if dtype in READ_FUNCS:
-        return READ_FUNCS[dtype](fi, length)
-    else:
-        data = np.fromstring(fi.read(int(length)), dtype)
-        if new_dtype is not None:  # cast data to new_dtype (due to #2198)
-            data = data.astype(new_dtype)
+    if dtype == 'bcd':
+        return _read_bcd(fi, length, left_part)
+    if dtype == 'binary':
+        return _read_binary(fi, length, left_part)
+    if dtype == 'IEEE':
+        data = np.frombuffer(fi.read(int(length)), '>f4')
         return data[0] if len(data) == 1 else data
 
 
-@_register_read_func('bcd')
-def _read_bcd(fi, length):
+def _read_bcd(fi, length, left_part):
```

Comment:
we have some BCD code in `io.reftek` already I think. Just leave as is for now, but ideally we want to avoid duplicated code..

Reply:
There was some talk on refactoring all of our parsers to use common utilities functions. It would  be worthwhile but a huge effort best left for another PR.

Reply:
I agree, should be done in another PR. Mostly mentioned it for future reference and FYI  :+1: